cluster_experiments/experiment_analysis.py

```diff
@@ -680,3 +683,186 @@ def analysis_point_estimate(self, df: pd.DataFrame, verbose: bool = False) -> fl
         results_mlm = self.fit_mlm(df)
         return results_mlm.params[self.treatment_col]
+
+
+class SyntheticControlAnalysis(ExperimentAnalysis):
+    """
+    Class to run Synthetic control analysis. It expects only one treatment cluster.
+
+    Arguments:
+
+        target_col (str): The name of the column containing the variable to measure.
+        treatment_col (str): The name of the column containing the treatment variable.
+        treatment (str): The name of the treatment to use as the treated group.
+        cluster_cols (list): A list of columns to use as clusters.
+        hypothesis (str): One of "two-sided", "less", "greater" indicating the hypothesis.
+        time_col (str): The name of the column containing the time data.
+        transition_date (str): The date when the intervention occurred.
+    Usage:
+
+    ```python
+    from cluster_experiments.experiment_analysis import SyntheticControlAnalysis
+    import pandas as pd
+    import numpy as np
+
+    dates = pd.date_range(start_date, end_date, freq="d")
+
+    users = [f"User {i}" for i in range(N)]
+
+    # Create a combination of each date with each user
+    combinations = list(product(users, dates))
+
+    target_values = np.random.normal(0, 1, size=len(combinations))
+
+    df = pd.DataFrame(combinations, columns=["user", "date"])
+    df["target"] = target_values
+
+    df["treatment_period"] = np.where(
+        df["date"].dt.date < date(2022, 1, 15), "Before", "After"
+    )
+
+    SyntheticControlAnalysis(
+        cluster_cols=["user"], time_col="date", transition_date=date(2022, 1, 15)
+    ).get_pvalue(df)
+
+    ```
+    """
+
+    def __init__(
+        self,
+        target_col: str = "target",
+        treatment_col: str = "treatment",
+        treatment: str = "B",
+        cluster_cols: Optional[List[str]] = None,
+        hypothesis: str = "two-sided",
+        time_col: str = "date",
+        transition_date: str = None,
+    ):
+        super().__init__(cluster_cols)
+
+        self.target_col = target_col
+        self.treatment = treatment
+        self.treatment_col = treatment_col
+        self.cluster_cols = cluster_cols
+        self.hypothesis = hypothesis
+        self.time_col = time_col
+        self.transition_date = transition_date
+
+    def fit_synthetic(self, df: pd.DataFrame) -> list:
+        """Returns the fitted OLS model"""
+
+        # todo should I throw an error if more than one cluster column is passed?
+
+        X = (
+            df.query(f"{self.treatment_col}==0")
+            .pivot(index=self.cluster_cols, columns=self.time_col)[self.target_col]
+            .T
+        )
+
+        if not any(df[self.treatment_col] == 1):
+            raise ValueError("No treatment unit found in the data.")
+
+        y = df.query(f"{self.treatment_col}==1")[f"{self.target_col}"]  # treated
+
+        weights = get_w(X, y)
+
+        return weights
+
+    def synthetic_control(self, df: pd.DataFrame, treatment_unit) -> np.array:
```

Comment:
what does this function actually do? I think it can be renamed to something more related to what it does

Reply:
and I think it returns a dataframe, not an array